src/main/java/org/jenkinsci/plugins/junitrealtimetestreporter/RealtimeJUnitStep.java

```diff
@@ -0,0 +1,217 @@
+    public void setTestDataPublishers(@Nonnull List<TestDataPublisher> testDataPublishers) {
+        this.testDataPublishers = testDataPublishers;
+    }
+
+    public double getHealthScaleFactor() {
+        return healthScaleFactor == null ? 1.0 : healthScaleFactor;
+    }
+
+    @DataBoundSetter
+    public void setHealthScaleFactor(double healthScaleFactor) {
+        this.healthScaleFactor = Math.max(0.0, healthScaleFactor);
+    }
+
+    public boolean isAllowEmptyResults() {
+        return allowEmptyResults;
+    }
+
+    @DataBoundSetter
+    public void setAllowEmptyResults(boolean allowEmptyResults) {
+        this.allowEmptyResults = allowEmptyResults;
+    }
+
+    @Override
+    public StepExecution start(StepContext context) throws Exception {
+        JUnitResultArchiver delegate = new JUnitResultArchiver(testResults);
+        delegate.setAllowEmptyResults(allowEmptyResults);
+        delegate.setHealthScaleFactor(getHealthScaleFactor());
+        delegate.setKeepLongStdio(keepLongStdio);
+        delegate.setTestDataPublishers(getTestDataPublishers());
+        return new Execution(context, delegate);
+    }
+
+    static class Execution extends StepExecution {
+
+        private final JUnitResultArchiver archiver;
+
+        Execution(StepContext context, JUnitResultArchiver archiver) {
+            super(context);
+            this.archiver = archiver;
+        }
+
+        @Override
+        public boolean start() throws Exception {
+            Run<?, ?> r = getContext().get(Run.class);
+            String id = getContext().get(FlowNode.class).getId();
+            r.addAction(new PipelineRealtimeTestResultAction(id, getContext().get(FilePath.class), archiver.isKeepLongStdio(), archiver.getTestResults()));
+            AbstractRealtimeTestResultAction.saveBuild(r);
+            getContext().newBodyInvoker().withCallback(new Callback(id, archiver)).start();
+            return false;
+        }
+
+        private static final long serialVersionUID = 1;
+
+    }
+
+    static class Callback extends BodyExecutionCallback.TailCall {
+
+        private final String id;
+        private final JUnitResultArchiver archiver;
+
+        Callback(String id, JUnitResultArchiver archiver) {
+            this.id = id;
+            this.archiver = archiver;
+        }
+
+        @Override
+        protected void finished(StepContext context) throws Exception {
+            Run<?, ?> r = context.get(Run.class);
+            for (PipelineRealtimeTestResultAction a : r.getActions(PipelineRealtimeTestResultAction.class)) {
+                if (a.id.equals(id)) {
+                    r.removeAction(a);
+                    LOGGER.log(Level.FINE, "clearing {0} from {1}", new Object[] {id, r});
+                    AbstractRealtimeTestResultAction.saveBuild(r);
+                    break;
+                }
+            }
+            archiver.perform(r, context.get(FilePath.class), context.get(Launcher.class), context.get(TaskListener.class));
```

Comment:
Note: might block CPS VM thread. Not trivial to solve: [JENKINS-43276](https://issues.jenkins-ci.org/browse/JENKINS-43276)